common-ts/src/lst/lst.ts

```diff
@@ -0,0 +1,236 @@
+ * @param lst 
+ * @param timestamps 
+ * @returns 
+ */
+export const getLstPriceMap = async (lst: LST, timestamps: number[]) => {
+	
+	const targetTimestampSet = new Set<number>(timestamps);
+	const descSortedPrices = (await getPrices(getDatasourceForLst(lst))).sort((a, b) => b.timestamp - a.timestamp);
+	const priceMap = new Map<number, number>();
+
+	// Need to match and add prices to the price map
+	targetTimestampSet.forEach((targetTimestamp) => {
+		// The matching price is the first price that is less than or equal to the target timestamp when going through the price entries in descending order .. aka the most recent price measurement for the target timestamp
+		const match = descSortedPrices.find((price) => price.timestamp <= targetTimestamp);
+		if (match) {
+			priceMap.set(targetTimestamp, match.price);
+		}
+	});	
+
+	return priceMap;
+};
+
+/**
+ * Returns a timestamp in seconds, for the start of the day in UTC format .. strips the hours, minutes and seconds
+ * @param timestamp 
+ */
+const getNormalisedTimestamp = (timestamp: number) => {
+	const date = new Date(Math.round(toMsecs(timestamp)));
+	date.setUTCHours(0, 0, 0, 0);
+	return date.getTime() / 1000;
+};
+
+export async function calculateSolEarned({
+	marketIndex,
+	user,
+	depositRecords,
+}: {
+	marketIndex: number;
+	user: User;
+	depositRecords: DepositRecord[];
+}): Promise<BN> {
+	const normalisedTimestamps: number[] = [
+		Date.now(),
+		...depositRecords
+			.filter((r) => r.marketIndex === marketIndex)
+			.map((r) => ((r.ts.toNumber()))),
+	].map(getNormalisedTimestamp);
+
+	const lst : LST = marketIndex === 2 ? 'msol' : marketIndex === 6 ? 'jitosol' : 'bsol';
+
+	const lstRatios = await getLstPriceMap(lst, normalisedTimestamps);
+
+	let solEarned = ZERO;
+	for (const record of depositRecords) {
+		if (record.marketIndex === 1) {
+			if (isVariant(record.direction, 'deposit')) {
+				solEarned = solEarned.sub(record.amount);
+			} else {
+				solEarned = solEarned.add(record.amount);
+			}
+		} else if (
+			record.marketIndex === 2 ||
+			record.marketIndex === 6 ||
+			record.marketIndex === 8
+		) {
+			const normalisedTimestamp = getNormalisedTimestamp(record.ts.toNumber());
+			const lstRatio = lstRatios.get(normalisedTimestamp);
+			const lstRatioBN = new BN(lstRatio * LAMPORTS_PER_SOL);
+
+			const solAmount = record.amount.mul(lstRatioBN).div(LAMPORTS_PRECISION);
+			if (isVariant(record.direction, 'deposit')) {
+				solEarned = solEarned.sub(solAmount);
+			} else {
+				solEarned = solEarned.add(solAmount);
+			}
+		}
+	}
+
+	const currentLstTokenAmount = await user.getTokenAmount(marketIndex);
+	const currentLstRatio = await getLstSolPrice(lst);
+	const currentLstRatioBN = new BN(currentLstRatio * LAMPORTS_PER_SOL);
+
+	solEarned = solEarned.add(
+		currentLstTokenAmount.mul(currentLstRatioBN).div(LAMPORTS_PRECISION)
+	);
```

Comment:
Also not strictly specific to your PR, but I was also wondering if we should do the "sol earned" feature differently from a UX point of view.

I had a friend sort of get confused about what that even means,  because there is no breakdown about where the sol earned is coming from or where it ends up. He thought he was getting sol that was claimable or added to his drift account or something lol.

I think we should work towards not showing a "sol earned" number necessarily, and instead showing a performance comparison of the user's current account value against the value of their initial deposit. That way it would factor out the "sol earned" that would have already been there if they had not been super staking at all.